pkg/server/router.go

```diff
@@ -293,11 +272,94 @@ func serveStatic(urlPrefix, staticDir string) http.Handler {
 	code := http.StatusInternalServerError
-	if len(status) == 1 {
+	if len(status) == 1 && status[0] != 0 {
 		code = status[0]
 	}
-
+	if err == nil {
+		err = errors.New("unknown")
+	}
 	logrus.Error(err)
 	resp.WriteHeader(code)
 	resp.Write([]byte(err.Error()))
 }
+
+// nodePassBootstrapper returns a node name, http code, and error
+type nodePassBootstrapper func(req *http.Request) (string, int, error)
+
+func passwordBootstrap(ctx context.Context, config *Config) nodePassBootstrapper {
+	runtime := config.ControlConfig.Runtime
+	var secretClient coreclient.SecretClient
+	var once sync.Once
+
+	return nodePassBootstrapper(func(req *http.Request) (string, int, error) {
+		nodeName, nodePassword, err := getNodeInfo(req)
+		if err != nil {
+			return "", http.StatusBadRequest, err
+		}
+
+		if secretClient == nil {
+			if runtime.Core != nil {
+				// initialize the client if we can
+				secretClient = runtime.Core.Core().V1().Secret()
+			} else if nodeName == os.Getenv("NODE_NAME") {
+				// or verify the password locally and ensure a secret later
+				return verifyLocalPassword(ctx, config, &once, nodeName, nodePassword)
+			} else {
+				// or reject the request until the core is ready
+				return "", http.StatusServiceUnavailable, errors.New("runtime core not ready")
+			}
+		}
+
+		if err := nodepassword.Ensure(secretClient, nodeName, nodePassword); err != nil {
+			return "", http.StatusForbidden, err
+		}
+
+		return nodeName, 0, nil
+	})
+}
+
+func verifyLocalPassword(ctx context.Context, config *Config, once *sync.Once, nodeName, nodePassword string) (string, int, error) {
+	// use same password file location that the agent creates
+	nodePasswordRoot := "/"
+	if config.Rootless {
+		nodePasswordRoot = filepath.Join(config.ControlConfig.DataDir, "agent")
+	}
+	nodeConfigPath := filepath.Join(nodePasswordRoot, "etc", "rancher", "node")
+	nodePasswordFile := filepath.Join(nodeConfigPath, "password")
+
+	passBytes, err := ioutil.ReadFile(nodePasswordFile)
+	if err != nil {
+		return "", http.StatusInternalServerError, errors.Wrap(err, "unabled to read node password file")
+	}
+
+	password := strings.TrimSpace(string(passBytes))
+	if password != nodePassword {
+		return "", http.StatusForbidden, errors.Wrapf(err, "unable to verify local password for node '%s'", nodeName)
+	}
+
+	// make sure the secret is created when the api server is ready
+	ensureSecret := func() {
+		runtime := config.ControlConfig.Runtime
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(1 * time.Second):
+				if runtime.Core != nil {
+					logrus.Debugf("runtime core has become available, ensuring password secret for node '%s'", nodeName)
+					secretClient := runtime.Core.Core().V1().Secret()
+					if err := nodepassword.Ensure(secretClient, nodeName, nodePassword); err != nil {
+						logrus.Warnf("error ensuring node password secret for pre-validated node '%s': %v", nodeName, err)
```

Comment:
1. Ensure will retry once if the resource has been created and it tries to create one.
2. Per the assumption made in question one, yes.
3. The implication is that a valid server/agent cert has been issued but does not match the secret stored in k8s. 
There is probably more work that needs to be done to make sure HA servers with duplicate hostnames can not exist. This will probably involve uplifting some of the nodeName and node password code for server bootstrapping.